pkg/cloudprovider/providers/vsphere/vclib/virtualmachine.go

```diff
@@ -0,0 +1,386 @@
+package vclib
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+
+	"github.com/golang/glog"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+// VirtualMachine extends the govmomi VirtualMachine object
+type VirtualMachine struct {
+	*object.VirtualMachine
+	datacenter *Datacenter
+}
+
+// IsDiskAttached checks if disk is attached to the VM.
+func (vm *VirtualMachine) IsDiskAttached(ctx context.Context, diskPath string) (bool, error) {
+	// Get object key of controller to which disk is attached
+	_, err := vm.GetVirtualDiskControllerKey(ctx, diskPath)
+	if err != nil {
+		if err == ErrNoDevicesFound {
+			return false, nil
+		}
+		glog.Errorf("Failed to check whether disk is attached to VM: %q. err: %s", vm.Name(), err)
+		return false, err
+	}
+	return true, nil
+}
+
+// GetVirtualDiskUUIDByPath gets the virtual disk UUID by diskPath
+func (vm *VirtualMachine) GetVirtualDiskUUIDByPath(ctx context.Context, diskPath string) (string, error) {
```

Comment:
The UUID this function returns is the virtual disk SCSI inquiry page 0x83 data.
This UUID will not be visible in vsphere management APIs anywhere.
Can we rename this function to "GetVirtualDiskPage83Data"? It will be more accurate.